_source/_posts/2022-06-16-oauth-java.adoc

```diff
@@ -0,0 +1,389 @@
+----
+
+====
+What annoys me about this configuration is `issuer` is an OpenID Connect concept that allows you to find all the endpoints for an identity provider, including the location of the JWKS keys. If you go to `${issuer}/.well-known/openid-configuration`, it has the location of the public keys in the `jwks_uri` value!
+
+image::{% asset_path 'blog/java-oauth2-2022/jwks-uri.png' %}[alt=The publickey.location can be calculated!,width=800,align=center]
+
+I really wish MicroProfile made things easier on the developer so you only have to specify the issuer.
+
+The good news is you should be able to refer to one property from another.
+
+[source,properties]
+----
+mp.jwt.verify.issuer=https://dev-13337.okta.com/oauth2/default
+mp.jwt.verify.publickey.location=${mp.jwt.verify.issuer}/v1/keys
+----
+
+*&lt;rant/&gt;*
+====
+
+==== Quarkus
+
+If you want to use MicroProfile JWT authentication with Quarkus, you need to add the properties above and a single dependency.
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-smallrye-jwt</artifactId>
+</dependency>
+----
+
+==== Helidon
+
+Helidon is very similar to Quarkus. The only difference is its dependency's coordinates.
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.helidon.microprofile.jwt</groupId>
+    <artifactId>helidon-microprofile-jwt-auth</artifactId>
+</dependency>
+----
+
+I've already mentioned Helidon and Quarkus, what about the other popular Java API frameworks?
+
+=== Micronaut
+
+Micronaut has support for OAuth and good documentation on how to configure it with IdPs like https://guides.micronaut.io/latest/micronaut-oauth2-okta.html[Okta], https://guides.micronaut.io/latest/micronaut-oauth2-oidc-google.html[Google], and https://guides.micronaut.io/latest/micronaut-oauth2-github.html[GitHub].
+
+Micronaut requires a lot more lines of configuration than MicroProfile, but it's mostly YAML's fault. It takes you 11 lines to configure an OAuth 2.0 resource server.
+
+[source,yaml]
+----
+micronaut:
+  security:
+    enabled: true
+    token:
+      jwt:
+        enabled: true
+        claims-validators:
+          issuer: https://dev-13337.okta.com/oauth2/default
+        signatures:
+          jwks:
+            okta:
+              url: https://dev-13337.okta.com/oauth2/default/v1/keys
+----
+
+As you can see, Micronaut also requires you to specify the JWKS location, even though it can be looked up from the issuer. `:shakes-fist:`
+
+=== Spring Boot
+
+Spring Boot uses Spring Security by default and it's the only framework that allows you to configure a resource server with one line of code.
+
+[source,properties]
+----
+spring.security.oauth2.resourceserver.jwt.issuer-uri=https://dev-13337.okta.com/oauth2/default
+----
+
+Of course, you could use YAML instead. That would be a bad life choice. Then it would require six lines of code. Don't be that person. Spring Security is smart enough to look up the JWKS location from the issuer.
```

Comment:
I think you're being cheeky here... would it make sense to add a winky emoji for the uninitiated? Just a thought...

Reply:
Added 😉